typed-transitions/src/Protocol/Chain/Type.hs

```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+
+{-# OPTIONS_GHC "-fwarn-incomplete-patterns" #-}
+
+module Protocol.Chain.Type where
+
+import Data.List.NonEmpty
+import Protocol.Core
+
+-- |
+-- = Definition of the chain exchange protocol, and a factoring of it into
+-- consumer and producer sides.
+
+-- | States in the chain exchange system.
+data StChainExchange where
+  -- | Initial state: the server sends two points.
+  StInit ::                StChainExchange
+  -- | Idle state: the client makes a choice.
+  StIdle ::                StChainExchange
+  -- | Busy state: the server responds to client choice.
+  StBusy :: RequestKind -> StChainExchange
+  -- | Done state: the client or the server terminated.
+  StDone ::                StChainExchange
+
+-- | Transitions on 'StChainExchange'. A request goes from idle to busy, and
+-- a response may go from busy to idle, or stay on busy in case of multi-part
+-- response for headers.
+data TrChainExchange point header stFrom stTo where
+  -- | The first transition gives the endpoints of the chain.
+  TrInit         :: (point, point)                -> TrChainExchange point header 'StInit         'StIdle
+  -- | A client request goes from 'StIdle to 'StBusy.
+  TrRequest      :: Request point req             -> TrChainExchange point header 'StIdle         ('StBusy req)
+  -- | A server response can go from 'StBusy to 'StIdle, or remain 'StBusy
+  -- in case of a multi-part header download.
+  TrRespond      :: Response point header req res -> TrChainExchange point header ('StBusy req)   res
+  -- | The client stops.
+  TrConsumerDone ::                                  TrChainExchange point header 'StIdle         'StDone
+  -- | The server stops.
+  TrProducerDone ::                                  TrChainExchange point header ('StBusy 'Next) 'StDone
+
+showTrChainExchange :: (Show point, Show header) => TrChainExchange point header from to -> String
+showTrChainExchange tr = case tr of
+  TrInit _ -> "Init"
+  TrRequest (ReqImprove hhs) -> "ReqSetHead " ++ show hhs
+  TrRequest (ReqDownload _) -> "ReqDownload"
+  TrRequest ReqNext -> "ReqNext"
+  TrRespond (ResImprove (hh, _)) -> "ResSetHead " ++ show hh
+  TrRespond (ResForked _) -> "ResChange"
+  TrRespond (ResDownloadOne _) -> "ResDownloadOne"
+  TrRespond (ResDownloadDone _) -> "ResDownloadDone"
+  TrRespond _ -> "Res other"
+  TrConsumerDone -> "Consumer done"
+  TrProducerDone -> "Producer done"
+
+data RequestKind where
+  Improve  :: RequestKind
+  Download :: RequestKind
+  -- | Request for the next change in the chain.
+  Next     :: RequestKind
+
+data Request point (req :: RequestKind) where
+  -- | Request to update the read pointer to the newest of these blocks.
+  ReqImprove  :: NonEmpty point -> Request point 'Improve
+  -- | Request at most this many headers, from the current read pointer.
+  ReqDownload :: Word           -> Request point 'Download
```

Comment:
`Word` could be changed for `Word32` or maybe even `Word16` since there is a limit how many blocks/headers can be downloaded imposed by other layers.  I remember @dcoutts mentioned this shortly.

Reply:
Should we include an option for "unlimited"?